src/lssi.c

```diff
@@ -0,0 +1,374 @@
+/* Lookahead sensative state item searches for counterexample generation
+ 
+ Copyright (C) 2019-2020 Free Software Foundation, Inc.
+ 
+ This file is part of Bison, the GNU Compiler Compiler.
+ 
+ This program is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+ 
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+ 
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.  */
+
+#include <config.h>
+#include <gl_linked_list.h>
+#include <gl_xlist.h>
+#include <stdlib.h>
+#include "lssi.h"
+#include "state-item.h"
+#include "nullable.h"
+#include "getargs.h"
+
+/** If set to true, when computing the shortest lookahead-sensitive path,
+ * only consider states that can reach the conflict state. */
+#define OPTIMIZE_SHORTEST_PATH true
+
+// lookahead sensative state item
+struct lssi;
+typedef struct lssi
+{
+  state_item_number si;
+  struct lssi *parent;
+  // this is the precise lookahead set (follow_L from the CupEx paper)
+  bitset lookahead;
+  bool free_lookahead;
+} lssi;
+
+lssi *
+new_lssi (state_item_number si, lssi *p, bitset l, bool free_l)
+{
+  lssi *ret = xmalloc (sizeof (lssi));
+  ret->si = si;
+  ret->parent = p;
+  ret->lookahead = l;
+  ret->free_lookahead = free_l;
+  return ret;
+}
+
+void
+lssi_free (lssi *sn)
+{
+  if (sn == NULL)
+    return;
+  if (sn->free_lookahead)
+    bitset_free (sn->lookahead);
+  free (sn);
+}
+
+static size_t
+lssi_hasher (lssi *sn, size_t max)
+{
+  size_t hash = sn->si;
+  bitset_iterator biter;
+  symbol_number syn;
+  BITSET_FOR_EACH (biter, sn->lookahead, syn, 0)
+    hash += syn;
+  return hash % max;
+}
+
+static bool
+lssi_comparator (lssi *s1, lssi *s2)
+{
+  if (s1->si == s2->si)
+    {
+      if (s1->lookahead == s2->lookahead)
+        return true;
+      return bitset_equal_p (s1->lookahead, s2->lookahead);
+    }
+  return false;
+}
```

Comment:
I like better mathematical style
``` C
return (s1->si == s2->si
        && (s1->lookahead == s2->lookahead
            || bitset_equal_p (s1->lookahead, s2->lookahead))
```
Actually, it should be
``` C
return s1->si == s2->si && bitset_equal_p (s1->lookahead, s2->lookahead);
```
and if there's really a diff in perf, it should be addressed in `bitset_equal_p` directly.

We'll see that after the merge.